README.txt
Perf4CDI - Perf4j integration for CDI (JSR-299)
Copyright (C) 2010 Marcin Zajaczkowski
Email: mszpak ATT wp DOTT pl
Homepage: http://perf4cdi.sourceforge.net/

Perf4CDI is available under the terms of GNU Lesser General Public License version 2.1 or (at your opinion) any later version.


Perf4CDI is an add-on package for implementations of Contexts and Dependency Injection for the Java EE platform (CDI - JSR-299) providing convenient integration with Perf4j - performance monitoring library for Java applications.
Perf4CDI was confirmed to work with OpenWebBeans 1.0.0-M4.
Due to a problem with Weld 1.0.1 final (see https://jira.jboss.org/browse/WELD-507) Perf4CDI should be able to work with upcomming Weld 1.0.2.CR1.


Those things are already available and ready to use:
 - a convenient integration with Perf4j - by annotating methods
 - a simple stop watch mechanism for succinct timing statements (Perf4j's feature)
 - a command line tool for parsing log files that generates aggregated statistics and performance graphs (Perf4j's feature)
 - an ability to expose performance statistics as JMX attributes, and to send notifications when statistics exceed specified thresholds (Perf4j's feature)
 - a servlet for exposing performance graphs in a web application (Perf4j's feature)

Things planned to implement in the nearest future:
 - CDI factory to produce loggers
 - ability to work as CDI extension (needed?)
 - move project's SCM to SourceForge - DONE
 - add unit tests - DONE
 - put project's jars into public Maven repository - DONE
 - further suggestions are welcome

Note: Perf4CDI in current version does its job, but there are a few things to improve, so it is not possible to quarantine a backward compatibility of the future versions. Therefore it should be used in production environmental carefully.


The best way to get support is the forum:
http://sourceforge.net/projects/seam-perf4j/forums/forum/1092214

Bugs can be reported using project's issue tracker:
http://sourceforge.net/tracker/?group_id=304978&atid=1285292


===================


Getting started guide

Maven configuration

For Maven projects it's required to add perf4cdi dependency to application's pom.xml:
<dependency>
    <groupId>net.sf.perf4cdi</groupId>
    <artifactId>perf4cdi</artifactId>
    <version>0.4.0</version>
</dependency>

Because Perf4CDI was created for CDI based applications it is assumed that cdi-api (tested with 1.0-SP1) is provided by runtime environment.
Perf4CDI internally uses slf4j API for logging. slf4j-api (tested with 1.6.0) should be provided as well.


CDI configuration

It is required to register additional interceptor 'net.sf.perf4cdi.Perf4jProfiledInterceptor' in beans.xml in CDI application to profile:

<beans>
    (...)
    <interceptors>
        <class>net.sf.perf4cdi.Perf4jProfiledInterceptor</class>
    </interceptors>
    (...)
</beans>


Application configuration

The last step is to annotate profiled method:
@Perf4jProfiled
@Profiled(tag = "profiledMethodTag")
public void profiledMethod() {...}

Note: Due to CDI architecture it's required to use marker @Perf4jProfiled annotation to associate interceptor with a method in addition to original @Profiled annotation from Perf4j. @Profiled annotation can be omitted (logging with default values takes place then).

After execution of profiled method(s) you should be able to see message similar to following in the application log:
INFO  [org.perf4j.TimingLogger] start[1267557706784] time[1075] tag[profiledMethodTag]

Happy profiling with CDI and Perf4j!


===================

Developer's guide - only for contributors

Perf4CDI itself depends only on packages which are available in Maven Central Repository. For test purposes however it uses mechanisms from weld-core-test which depends on JARs available only in JBoss' Maven Repository. Because of that to build Perf4CDI from source (only contributors would need to do that - for normal usage in end applications Maven Central Repository should be enough) it's required to add repository in settings.xml located in user's home directory:

<repositories>
    (... - some other repositories)
    <repository>
        <id>jboss</id>
        <url>http://repository.jboss.com/maven2</url>
        <releases>
            <enabled>true<enabled>
        </releases>
        <snapshots>
            <enabled>false</enabled>
        </snapshots>
    </repository>
</repositories>



CHANGELOG.txt
Perf4CDI changelog.

0.4.0 (2010-07-28):
 - added mock unit tests
 - availability in Maven Cetral Repository

0.3.0 (2010-07-02):
 - initial public release


src/test/java/net/sf/perf4cdi/Perf4jProfiledTestBean.java
package net.sf.perf4cdi;

/**
 * Interface for Perf4j profiled test bean.
 * 
 * Common part for Seam-Perf4j and Perf4CDI.
 *
 * @author Marcin Zajączkowski, 2010-03-22
 */
public interface Perf4jProfiledTestBean {

    public static final String PROFILED_METHOD_NAME = "profiledMethod";
    
    public long profiledMethod(long sleepTime) throws Exception;
}


src/test/java/net/sf/perf4cdi/CDIInMemoryTimingAspect.java
package net.sf.perf4cdi;

import org.perf4j.LoggingStopWatch;

import java.util.ArrayList;
import java.util.List;

/**
 * Simple in memory timing aspect used in tests.
 *
 * Common part for Seam-Perf4j and Perf4CDI.
 *
 * @author Marcin Zajączkowski, 2010-03-14
 */
public class CDIInMemoryTimingAspect extends CDITimingAspect {

    private List<String> loggedMessages = new ArrayList<String>();
    
    @Override
    protected LoggingStopWatch newStopWatch(String loggerName, String levelName) {

        return new LoggingStopWatch() {
            @Override
            public boolean isLogging() {
                return true;
            }

            @Override
            protected void log(String stopWatchAsString, Throwable exception) {
                loggedMessages.add(stopWatchAsString);
            }
        };
    }

    public String getLastLoggedMessage() {
        if (!loggedMessages.isEmpty()) {
            return loggedMessages.get(loggedMessages.size() - 1);
        } else {
            return null;
        }
    }
}


src/test/java/net/sf/perf4cdi/AbstractPerf4jProfiledTestBean.java
package net.sf.perf4cdi;

import net.sf.perf4cdi.api.Perf4jProfiled;
import org.perf4j.aop.Profiled;

import javax.inject.Named;

/**
 * Abstract super class for ProfiledTestBeans
 *
 * Common part for Seam-Perf4j and Perf4CDI.
 *
 * @author Marcin Zajączkowski, 2010-03-22
 */
public abstract class AbstractPerf4jProfiledTestBean implements Perf4jProfiledTestBean {

    public long profiledMethod(long sleepTime) throws Exception {
        Thread.sleep(sleepTime);
        return sleepTime;
    }
}

/**
 * org.sf.seam.perf4j.SimplePerf4jProfiledTestBean
 *
 * @author Marcin Zajączkowski, 2010-03-22
 */
@Named("profiledWithCustomTag")
class CustomTagPerf4jProfiledTestBean extends AbstractPerf4jProfiledTestBean {

    public static final String CUSTOM_TAG = "profiledWithCustomTagTag";

    @Profiled(tag = CUSTOM_TAG)
    @Perf4jProfiled
    public long profiledMethod(long sleepTime) throws Exception {
        return super.profiledMethod(sleepTime);
    }
}

/**
 * org.sf.seam.perf4j.SimplePerf4jProfiledTestBean
 *
 * @author Marcin Zajączkowski, 2010-03-22
 */
@Named("simpleProfiled")
class SimplePerf4jProfiledTestBean extends AbstractPerf4jProfiledTestBean {

    @Profiled
    @Perf4jProfiled
    public long profiledMethod(long sleepTime) throws Exception {
        return super.profiledMethod(sleepTime);
    }
}

/**
 * org.sf.seam.perf4j.SimplePerf4jProfiledTestBean
 *
 * @author Marcin Zajączkowski, 2010-03-27
 */
@Named("profiledWithoutProfiledAnnotation")
class WithoutProfiledPerf4jProfiledTestBean extends AbstractPerf4jProfiledTestBean {

    @Perf4jProfiled
    public long profiledMethod(long sleepTime) throws Exception {
        return super.profiledMethod(sleepTime);
    }
}

/**
 * org.sf.seam.perf4j.SimplePerf4jProfiledTestBean
 *
 * @author Marcin Zajączkowski, 2010-03-27
 */
@Named("noProfiled")
class NoProfiledPerf4jProfiledTestBean extends AbstractPerf4jProfiledTestBean {

    public long profiledMethod(long sleepTime) throws Exception {
        return super.profiledMethod(sleepTime);
    }
}


src/main/java/net/sf/perf4cdi/CDIJoinPoint.java
/*
 * Perf4CDI - Perf4j integration for CDI (JSR-299)
 * Copyright (C) 2010 Marcin Zajaczkowski
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
package net.sf.perf4cdi;

import org.perf4j.aop.AbstractJoinPoint;

import javax.interceptor.InvocationContext;

/**
 * An implementation of AbstractJoinPoint from Perf4j library for CDI.
 *
 * CDIJoinPoint should be only created for method execution.
 * 
 * Note: that construction required at least version 0.9.13 of Perf4j.
 *
 * @author Marcin Zajączkowski, 2010-05-22
 */
public class CDIJoinPoint implements AbstractJoinPoint {

    private final InvocationContext ic;

    /**
     * Constructor with CDI InvocationContext to wrap.
     *
     * @param ic original InvocationContext to wrap
     */
    public CDIJoinPoint(InvocationContext ic) {
        this.ic = ic;
    }

    /**
     * {@inheritDoc}
     */
    public Object proceed() throws Exception {
        return ic.proceed();
    }

    /**
     * {@inheritDoc}
     */
    public Object getExecutingObject() {
        return ic.getTarget();
    }

    /**
     * {@inheritDoc}
     */
    public Object[] getParameters() {
        return ic.getParameters();
    }

    /**
     * {@inheritDoc}
     */
    public String getMethodName() {
        return ic.getMethod().getName();
    }
}


src/main/java/net/sf/perf4cdi/api/Perf4jProfiled.java
/*
 * Perf4CDI - Perf4j integration for CDI (JSR-299)
 * Copyright (C) 2010 Marcin Zajaczkowski
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
package net.sf.perf4cdi.api;

import javax.interceptor.InterceptorBinding;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Marker annotation for methods profiled with Perf4j.
 *
 * Methods to be profiled should be marked with that annotation.
 * <p/>
 * Due to CDI architecture in addition to original Profiled annotation from Perf4j library (which for example allows
 * to configure logging format) profiled method has to marked using that annotation as well.
 * Profiled annotation can be omitted (logging with default values takes place then).
 *
 * ElementType.TYPE was forced by requirements for CDI interceptors definition. Perf4jProfiled annotation can be
 * only used with mothods and constructors
 *
 * @author Marcin Zajączkowski, 2010-05-15
 *
 * @see net.sf.perf4cdi.Perf4jProfiledInterceptor
 */
@Target({ElementType.METHOD, ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
@InterceptorBinding
public @interface Perf4jProfiled {
}


src/main/java/net/sf/perf4cdi/Perf4jProfiledInterceptor.java
/*
 * Perf4CDI - Perf4j integration for CDI (JSR-299)
 * Copyright (C) 2010 Marcin Zajaczkowski
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
package net.sf.perf4cdi;

import net.sf.perf4cdi.api.Perf4jProfiled;
import org.perf4j.aop.DefaultProfiled;
import org.perf4j.aop.Profiled;

import javax.interceptor.AroundInvoke;
import javax.interceptor.Interceptor;
import javax.interceptor.InvocationContext;
import java.lang.reflect.Method;

/**
 * Perf4j interceptor for CDI.
 *
 * Note: it's required to turn on that interceptor in beans.xml interceptors:class.
 *
 * @author Marcin Zajączkowski, 2010-05-17
 *
 * @see net.sf.perf4cdi.api.Perf4jProfiled
 */
@Interceptor
@Perf4jProfiled
public class Perf4jProfiledInterceptor {

    private static final long serialVersionUID = -8572318944020712128L;

    /**
     * Delegates performance logging to specific TimingAspect class for annotated methods.
     *
     * @param ic invocation context
     * @return method execution return value
     * @throws Exception exception thrown during excecution
     */
    @AroundInvoke
    public Object aroundInvoke(InvocationContext ic) throws Exception {

        Method method = ic.getMethod();
        if (shouldMethodBeSkipped(method)) {
            return ic.proceed();
        }

        Profiled profiled = getProperProfiledAnnotation(method);

        //TODO: MZA: Could be moved to separate method
        CDITimingAspect ata = newTimingAspect();
        try {
            return ata.doPerfLogging(new CDIJoinPoint(ic), profiled);
        } catch (Exception e) {
            throw e;
        } catch (Throwable throwable) {
            //construction forced by Throwable in AgnosticTimingAspect.runProfiledMethod() vs.
            //Exception in OptimizedInterceptor.aroundInvoke()
            throw new Exception("Throwable in AgnosticTimingAspect.runProfiledMethod()", throwable);
        }
    }

    private boolean shouldMethodBeSkipped(Method method) {
        //Perf4jProfiled can be used only with method or constructor, but due to CDI requirements Perf4jProfiled
        //annotation has to have also TYPE Target. Interceptor could be also used directly so it's better to
        //verify that there is a method to prevent NPE.
        return method == null || method.getAnnotation(Perf4jProfiled.class) == null;
    }

    private Profiled getProperProfiledAnnotation(Method method) {
        Profiled profiled = method.getAnnotation(Profiled.class);
        //when no Profiled annotation was used default values are used
        if (profiled == null) {
            profiled = DefaultProfiled.INSTANCE;
        }
        return profiled;
    }


    /**
     * Returns new TimingAspect.
     *
     * Made mostly for tests.
     *
     * @return new instance of CDITimingAspect
     */
    protected CDITimingAspect newTimingAspect() {
        return new CDITimingAspect();
    }
}


src/main/java/net/sf/perf4cdi/CDITimingAspect.java
/*
 * Perf4CDI - Perf4j integration for CDI (JSR-299)
 * Copyright (C) 2010 Marcin Zajaczkowski
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
package net.sf.perf4cdi;

import org.perf4j.LoggingStopWatch;
import org.perf4j.aop.AgnosticTimingAspect;
import org.perf4j.aop.Profiled;
import org.perf4j.slf4j.Slf4JStopWatch;
import org.slf4j.LoggerFactory;


/**
 * An implementation of Perf4j's AgnosticTimingAspect using SLF4J logger to log results.
 *
 * @author Marcin Zajączkowski, 2010-05-22
 */
public class CDITimingAspect extends AgnosticTimingAspect {

    /**
     * Executes the real profiling method from Perf4j.
     *
     * @param cjp CDI join point
     * @param profiled profiled annotation from profiled method
     * @return return value from executed methof
     * @throws Throwable exception thrown by executing method
     */
    public Object doPerfLogging(CDIJoinPoint cjp, Profiled profiled) throws Throwable {
        return runProfiledMethod(cjp, profiled, newStopWatch(profiled.logger(), profiled.level()));
    }

    /**
     * Returns an implementation of StopWatch using Slf4j logger.
     *
     * @param loggerName logger name passed to StopWatch
     * @param levelName logging level name passed to StopWatch
     * @return LoggingStopWatch implementation
     */
    protected LoggingStopWatch newStopWatch(String loggerName, String levelName) {
        //TODO: MZA: A CDI factory could be created to produce loggers 
        return new Slf4JStopWatch(LoggerFactory.getLogger(loggerName), Slf4JStopWatch.mapLevelName(levelName));
    }
}
